### PR TITLE
Add BackendServiceMetadataUpdate test

### PR DESCRIPTION
**Pull request by akwi-github**
Backend service metadata update test

**Setup**

1. Create HealthCheck
2. Create BackendService with multiple keys in Metadata field and add HealthCheck to it

**Changes**

1. Update BackendService Metadatas field by updating one of key-pair values
2. Delete one of keys from the Metadatas field

**Expectations**

Each time expect that the comparison works properly

### CODE REVIEW

File: e2e/backend_service_update_test.go [hunk 1]
```diff
@@ -226,3 +255,86 @@ func TestBackendServiceUpdate(t *testing.T) {
...
 }
+
+func TestBackendServiceMetadataUpdate(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	graphBuilder := rgraph.NewBuilder()
+	resUniqueIdPart := "test-bs-metadata-update"
+
+	hcName := resourceName(resUniqueIdPart + "-hc")
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:266]: I think that `buildHealthCheck` already use `resourceName` inside this function

File: e2e/backend_service_update_test.go [hunk 2]
```diff
@@ -226,3 +255,86 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+	hcID, err := buildHealthCheck(graphBuilder, hcName, 15)
+	if err != nil {
+		t.Fatalf("buildHealthCheck(_, %s, _) = (_, %v) want (_, nil)", hcName, err)
+	}
+
+	bsName := resourceName(resUniqueIdPart + "-bs")
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:272]: as above

File: e2e/backend_service_update_test.go [hunk 3]
```diff
@@ -226,3 +255,86 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+	bsID, err := buildBackendServiceWithMetadata(graphBuilder, bsName, hcID, map[string]string{"foo": "1", "bar": "2", "baz": "2"})
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:273]: use wantedMetadata variable (you need to move it up:))

File: e2e/backend_service_update_test.go [hunk 4]
```diff
@@ -226,3 +255,86 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+
+func TestBackendServiceMetadataUpdate(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	graphBuilder := rgraph.NewBuilder()
+	resUniqueIdPart := "test-bs-metadata-update"
+
+	hcName := resourceName(resUniqueIdPart + "-hc")
+	hcID, err := buildHealthCheck(graphBuilder, hcName, 15)
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:267]: I don't think that HC is needed in this test, it will just make test execute longer but it has no meaning to the logic.

File: e2e/backend_service_update_test.go [hunk 5]
```diff
@@ -226,3 +255,86 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+	bs, err = theCloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("theCloud.BackendServices().Get(_, %s) = (nil, %v) wanted (_, nil)", bsID.Key, err)
+	}
+
+	if df := cmp.Diff(bs.Metadatas, wantedMetadata); df != "" {
+		t.Fatalf("Metadata keys are not updated correctly(-got, +want): %s", df)
+	}
+
+	bsID, err = buildBackendServiceWithMetadata(graphBuilder, bsName, hcID, map[string]string{"foo": "1", "bar": "2"})
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:323]: pleas add comments to each update operation what is this updating like: delete key, update one key etc

File: e2e/backend_service_update_test.go [hunk 6]
```diff
@@ -46,6 +46,35 @@ func buildBackendServiceWithLBScheme(graphBuilder *rgraph.Builder, name string,
 	return bsID, nil
 }
 
+func buildBackendServiceWithMetadata(graphBuilder *rgraph.Builder, name string, hcID *cloud.ResourceID, metadata map[string]string) (*cloud.ResourceID, error) {
+	bsID := backendservice.ID(testFlags.project, meta.GlobalKey(resourceName(name)))
+
+	bsMutResource := backendservice.NewMutableBackendService(testFlags.project, bsID.Key)
+	bsMutResource.Access(func(x *compute.BackendService) {
```
[the remark this answers is not in the document]
Comment (akwi-github) [reply e2e/backend_service_update_test.go:53]: Done.

File: e2e/backend_service_update_test.go (hunk 1 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:266]: Done.

File: e2e/backend_service_update_test.go (hunk 2 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:272]: Done.

File: e2e/backend_service_update_test.go (hunk 3 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:273]: Done.

File: e2e/backend_service_update_test.go (hunk 4 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:267]: Done.

File: e2e/backend_service_update_test.go (hunk 5 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:323]: Done.

File: e2e/backend_service_update_test.go [hunk 7]
```diff
@@ -31,12 +28,19 @@ func buildBackendServiceWithLBScheme(graphBuilder *rgraph.Builder, name string,
 		x.TimeoutSec = 30
 		x.ConnectionDraining = &compute.ConnectionDraining{}
 		x.HealthChecks = []string{hcID.SelfLink(meta.VersionGA)}
-	})
+	}
+
+	return buildBackendServiceWith(graphBuilder, name, f)
+}
+
+func buildBackendServiceWith(graphBuilder *rgraph.Builder, name string, f func(x *compute.BackendService)) (*cloud.ResourceID, error) {
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:36]: `buildBackendServiceWith` -> `buildBackendService`

File: e2e/backend_service_update_test.go [hunk 8]
```diff
@@ -226,3 +244,76 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+	}
+
+	/* Edit BackendService metadatas field */
+	bsID, err = buildBackendServiceWithMetadata(graphBuilder, bsName, map[string]string{"foo": "1", "bar": "2", "baz": "3"})
+	expectedActions = []exec.ActionMetadata{
+		{Type: exec.ActionTypeUpdate, Name: actionName(exec.ActionTypeUpdate, bsID)},
+	}
+
+	processGraphAndExpectActions(t, graphBuilder, expectedActions)
+	wantedMetadata = map[string]string{"foo": "1", "bar": "2", "baz": "3"}
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:289]: move this to line 281 and pass to `buildBackendServiceWithMetadata` function

File: e2e/backend_service_update_test.go [hunk 9]
```diff
@@ -226,3 +244,76 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+
+	/* Verify Metadatas field was updated correctly */
+	bs, err = theCloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("theCloud.BackendServices().Get(_, %s) = (nil, %v) wanted (_, nil)", bsID.Key, err)
+	}
+
+	if df := cmp.Diff(bs.Metadatas, wantedMetadata); df != "" {
+		t.Fatalf("Metadata keys are not updated correctly(-got, +want): %s", df)
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:298]: please unify error messaged, the one in line 279 is different

File: e2e/backend_service_update_test.go [hunk 10]
```diff
@@ -226,3 +244,76 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+	}
+
+	/* Remove a key form BackendService metadatas field */
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:301]: form -> from

File: e2e/backend_service_update_test.go [hunk 11]
```diff
@@ -226,3 +244,76 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+	bsID, err = buildBackendServiceWithMetadata(graphBuilder, bsName, map[string]string{"foo": "1", "bar": "2"})
+	expectedActions = []exec.ActionMetadata{
+		{Type: exec.ActionTypeUpdate, Name: actionName(exec.ActionTypeUpdate, bsID)},
+	}
+
+	processGraphAndExpectActions(t, graphBuilder, expectedActions)
+	wantedMetadata = map[string]string{"foo": "1", "bar": "2"}
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:308]: as above use it as a function parameter

File: e2e/backend_service_update_test.go [hunk 12]
```diff
@@ -226,3 +244,76 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+
+	/* Verify a key form BackendService metadatas field was removed correctly */
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:310]: type: form -> from

File: e2e/backend_service_update_test.go (hunk 7 above)
Comment (kl52752) [reply e2e/backend_service_update_test.go:36]: there is also `buildBackendServiceWithNEG` please update this function as well

Comment (akwi-github) [reply e2e/backend_service_update_test.go:36]: there already exists a function buildBackendService, not renaming.

Comment (akwi-github) [reply e2e/backend_service_update_test.go:36]: Done.

File: e2e/backend_service_update_test.go (hunk 8 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:289]: Done.

File: e2e/backend_service_update_test.go (hunk 9 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:298]: The message in line 279 is a different message because it's about setting metadatas for the first time not updating them (even though technically it might be performed the same way I did it on purpose)

File: e2e/backend_service_update_test.go (hunk 10 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:301]: Done.

File: e2e/backend_service_update_test.go (hunk 12 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:310]: Done.

File: e2e/backend_service_update_test.go (hunk 9 above)
Comment (kl52752) [reply e2e/backend_service_update_test.go:299]: In line 279 (-got, + want) is missing, please at it

**Review by kl52752 — COMMENTED** [review]
One small nit, otherwise LGTM

File: e2e/backend_service_update_test.go (hunk 9 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:299]: Oh, I see. Done. Thanks.

**Comment by akwi-github** [comment]
> One small nit, otherwise LGTM

Done.

File: e2e/backend_service_update_test.go [hunk 13]
```diff
@@ -226,3 +244,78 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+
+	processGraphAndExpectActions(t, graphBuilder, expectedActions)
+
+	/* Verify Metadatas field was updated correctly */
+	bs, err = theCloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("theCloud.BackendServices().Get(_, %s) = (nil, %v) wanted (_, nil)", bsID.Key, err)
+	}
+
+	if df := cmp.Diff(bs.Metadatas, wantedMetadata); df != "" {
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:298]: use maps.Equal for check and cmp.Diff to print the diff if needed.

**Review by kl52752 — COMMENTED** [review]
one nit otherwise LGTM

File: e2e/backend_service_update_test.go (hunk 13 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go]: Done.

File: e2e/backend_service_update_test.go [hunk 14]
```diff
@@ -226,3 +245,78 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+
+	processGraphAndExpectActions(t, graphBuilder, expectedActions)
+
+	/* Verify Metadatas field was updated correctly */
+	bs, err = theCloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("theCloud.BackendServices().Get(_, %s) = (nil, %v) wanted (_, nil)", bsID.Key, err)
+	}
+
+	if df := cmp.Diff(bs.Metadatas, wantedMetadata); !maps.Equal(bs.Metadatas, wantedMetadata) {
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:299]: if !maps.Equal(bs.Metadatas, wantedMetadata) {
       df := cmp.Diff(bs.Metadatas, wantedMetadata); 
	t.Fatalf("Metadata keys are not updated correctly(-got, +want): %s", df)
}

File: e2e/backend_service_update_test.go [hunk 15]
```diff
@@ -227,6 +246,83 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+		}
+	})
+
+	/* Verify BackendService was created correctly */
+	bs, err := theCloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("theCloud.BackendServices().Get(_, %s) = (nil, %v) wanted (_, nil)", bsID.Key, err)
+	}
+
+	if df := cmp.Diff(bs.Metadatas, wantedMetadata); df != "" {
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:279]: use maps.Equal

File: e2e/backend_service_update_test.go [hunk 16]
```diff
@@ -227,6 +246,83 @@ func TestBackendServiceUpdate(t *testing.T) {
...
+
+	if !maps.Equal(bs.Metadatas, wantedMetadata) {
+		df := cmp.Diff(bs.Metadatas, wantedMetadata)
+		t.Fatalf("Metadata keys are not updated correctly(-got, +want): %s", df)
+	}
+
+	wantedMetadata = map[string]string{"foo": "1", "bar": "2"}
+
+	/* Remove a key from BackendService metadatas field */
+	bsID, err = buildBackendServiceWithMetadata(graphBuilder, bsName, wantedMetadata)
```
Comment (kl52752) [inline e2e/backend_service_update_test.go:307]: add another planing (without modifying metadatas) at the end end expect that no action is needed (this will verify how diff behave for maps)

File: go.mod [hunk 17]
```diff
@@ -1,6 +1,6 @@
 module github.com/GoogleCloudPlatform/k8s-cloud-provider
 
-go 1.21
```
Comment (kl52752) [inline go.mod:3]: do this change in separate PR

Comment (AwesomePatrol) [reply go.mod:3]: why? this is the first PR that requires a newer version and it doesn't mean anything besides "I need stuff from at least this version". PR for one-line change seems like an overkill

File: e2e/backend_service_update_test.go (hunk 15 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go]: Done.

File: e2e/backend_service_update_test.go (hunk 16 above)
Comment (akwi-github) [reply e2e/backend_service_update_test.go:307]: Done.

File: go.mod (hunk 17 above)
Comment (akwi-github) [reply go.mod:3]: If I'd do so, tests would fail.

Comment (kl52752) [reply go.mod:3]: tests are still failing, what needs 1.22. maps.Equal is in 1.21
https://screenshot.googleplex.com/5VEAYaUxvSdLLKQ.png

Comment (akwi-github) [reply go.mod]: I will try test the old version with the new image

Comment (AwesomePatrol) [reply go.mod]: I think you meant to show `1.21` not `1.22.1`. However, you are fully correct. Go version bump is not needed for `maps.Equal` (or anything else in this PR).

Comment (akwi-github) [reply go.mod]: Done.
